### ISSUE TITLE
Add tosa and/or linalg folders for transposes.

### ISSUE THREAD

**Issue by benvanik** [google/iree#7036]
ssd_mobilenet_v2_float has some transposes that are not folded into constants; this leads to us allocating the memory duplicating the constant and then transposing into it on every run. In this model this happens for every constant. I don't know why the transposes don't fold into other linalg ops, but we really shouldn't be doing them there anyway.

```mlir
%3 = "tosa.const"() {value = dense<"..."> : tensor<1x3x3x32xf32>} : () -> tensor<1x3x3x32xf32>
%145 = "tosa.const"() {value = dense<[1, 2, 3, 0]> : tensor<4xi32>} : () -> tensor<4xi32>
%150 = "tosa.transpose"(%3, %145) : (tensor<1x3x3x32xf32>, tensor<4xi32>) -> tensor<3x3x32x1xf32>
```
->
```mlir
%283 = linalg.generic {indexing_maps = [#map2, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%cst_74 : tensor<576xf32>) outs(%282 : tensor<1x20x20x576xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):  // no predecessors
      linalg.yield %arg1 : f32
    } -> tensor<1x20x20x576xf32>
```
->
```mlir
%167 = flow.dispatch @main_dispatch_81::@main_dispatch_81[%c384, %c64, %c1](%cst_74) : (tensor<384x64xf32>) -> tensor<64x384xf32>
```
->
```mlir
%279 = stream.async.dispatch @main_dispatch_240::@main_dispatch_240[%c256, %c128, %c3](%_constant_112) : (!stream.resource<constant>{%c1179648}) -> !stream.resource<transient>{%c1179648}
```

So, let's not do that :)
Having the ability to fold the linalg.generic into the constant would be amazing. Maybe this exists already but isn't wired up? Worst case we can fold the tosa op (easier), but linalg feels like where we need to be doing this (it'll show up a lot).

**Comment by asaadaldien** [comment]
AFIAK we don't have linalg constant folding, so I locally have a pass that constant folds many `linalg.generic` forms including transposes. IMO we can have it in Flow and tailor it to our cases because there are many compile time /runtime trade offs (e.g I am limiting the pass to only constant fold generic ops with `bytes(input_tensor) == bytes(output_tensor)`.

**Comment by antiagainst** [comment]
+1. Actually I was trying to do this in my previous attempt https://github.com/google/iree/pull/6614. I also mentioned adding a folder for transpose over constant filters when @rsuderman changes how upstream converts TOSA conv ops into Linalg ones https://reviews.llvm.org/D107504#inline-1023784. Rob said he will look into it but looks he hasn't gotten to it yet. @rsuderman any update on that? Doing it at Linalg level is also fine to me as that would handle cases from other inputs too.

**Comment by benvanik** [comment]
Yeah I think expanding is generally pretty difficult but it'd be nice to have some that did contraction/1:1 where we can safely do it. The TOSA level may be easiest as it's generally just a single function with the constants in it already deduplicated so checking for other uses such that you only are only folding if there's one user is easy. In more general linalg programs it's trickier as the same constant could be in multiple functions and folding in one but not another would be bad - but our constant outlining into globals helps: if we have a util.global of a constant value that is only ever loaded and then has a linalg op applied consistently at all load sites we can safely fold that in.
